File: Lab06_20221750/20221750/TestSoft/TestSoftDBManager/src/main/java/pe/edu/pucp/testsoft/db/DBFactoryProvider.java
package pe.edu.pucp.testsoft.db;

import pe.edu.pucp.testsoft.db.mysql.MySQLDBManagerFactory;
import pe.edu.pucp.testsoft.db.utils.TipoDB;

import java.util.ResourceBundle;

public class DBFactoryProvider {
    private static DBManager instancia;

    public static synchronized DBManager getManager() {
        if (instancia == null) {
            ResourceBundle properties = ResourceBundle.getBundle("db");

            String host = properties.getString("db.host");
            int puerto = Integer.parseInt(properties.getString("db.puerto"));
            String esquema = properties.getString("db.esquema");
            String usuario = properties.getString("db.usuario");
            String password = properties.getString("db.password");
            TipoDB tipo = TipoDB.valueOf(properties.getString("db.tipo"));

            DBManagerFactory factory;
            switch (tipo) {
                case MySQL -> factory = new MySQLDBManagerFactory();
                default -> throw new IllegalArgumentException("Tipo de DB no "
                        + "soportado: " + tipo);
            }

            instancia = factory.crearDBManager(host, puerto, esquema, usuario,
                    password);
        }

        return instancia;
    }
}

File: sesion-de-clase-15/SoftProg/SoftProgDBManager/Db/DbFactoryProvider.cs
using Microsoft.Extensions.Configuration;
using SoftProgDBManager.Db.Utils;

namespace SoftProgDBManager.Db;

public static class DbFactoryProvider
{
    private static DbManager? _instancia;

    public static DbManager GetManager()
    {
        if (_instancia is not null)
        {
            return _instancia;
        }

        var config = ConfigurationContext.Current;

        var tipoRaw = config["DbSettings:Tipo"]
            ?? "MySQL";
        var tipo = Enum.Parse<TipoDb>(tipoRaw, ignoreCase: true);

        var connectionStringBase = config.GetConnectionString("SoftProgDb")
            ?? throw new InvalidOperationException("No se encontro ConnectionStrings:SoftProgDb en la configuracion.");

        DbManagerFactory factory = tipo switch
        {
            TipoDb.MSSQL => new MsSqlDbManagerFactory(),
            TipoDb.MySQL => new MySqlDbManagerFactory(),
            _ => throw new ArgumentOutOfRangeException(nameof(tipo), tipo, "Tipo de DB no soportado")
        };

        _instancia = factory.CrearDbManager(connectionStringBase);
        return _instancia;
    }
}
